### PR TITLE
Daemon registry client

### PR DESCRIPTION
**Pull request by vishh**
Forked from #12383. cc @nikhiljindal.

### CODE REVIEW

File: pkg/client/unversioned/cache/listers.go [hunk 1]
```diff
@@ -225,6 +225,58 @@ func (s *StoreToReplicationControllerLister) GetPodControllers(pod *api.Pod) (co
...
+	}
+
+	for _, m := range s.Store.List() {
+		daemonController = *m.(*api.Daemon)
+		if daemonController.Namespace != pod.Namespace {
+			continue
+		}
+		selector = labels.Set(daemonController.Spec.Selector).AsSelector()
+
+		// If an daemonController with a nil or empty selector creeps in, it should match nothing, not everything.
```
Comment (nikhiljindal) [inline pkg/client/unversioned/cache/listers.go]: if a daemonController ....

File: pkg/client/unversioned/daemon.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+type DaemonInterface interface {
+	List(selector labels.Selector) (*api.DaemonList, error)
+	Get(name string) (*api.Daemon, error)
+	Create(ctrl *api.Daemon) (*api.Daemon, error)
+	Update(ctrl *api.Daemon) (*api.Daemon, error)
+	Delete(name string) error
+	Watch(label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error)
+}
+
+// daemons implements DaemonsNamespacer interface
```
Comment (nikhiljindal) [inline pkg/client/unversioned/daemon.go]: How about adding ```var _ DaemonsNamespacer = &daemons{}``` to ensure that this is true (and remains that way)

File: pkg/client/unversioned/testclient/fake_daemons.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+package testclient
+
+import (
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+// FakeDaemons implements DaemonInterface. Meant to be embedded into a struct to get a default
```
Comment (nikhiljindal) [inline pkg/client/unversioned/testclient/fake_daemons.go]: Same. How about adding var _ DaemonInterface = &FakeDaemons{}?

**Comment by nikhiljindal** [comment]
Mostly LGTM, just some minor nits.

File: pkg/client/unversioned/testclient/fake_daemons.go (hunk 3 above)
Comment (vishh) [inline pkg/client/unversioned/testclient/fake_daemons.go]: Done

File: pkg/client/unversioned/daemon.go (hunk 2 above)
Comment (vishh) [inline pkg/client/unversioned/daemon.go]: Done

File: pkg/client/unversioned/cache/listers.go (hunk 1 above)
Comment (vishh) [inline pkg/client/unversioned/cache/listers.go]: Goot catch. Fixed

**Comment by vishh** [comment]
@nikhiljindal: Addressed nits. PTAL.

**Comment by vishh** [comment]
@googlebot: @AnanyaKumar is ok with submitting his patches.

**Comment by nikhiljindal** [comment]
LGTM, thanks for the fixes!

**Comment by vishh** [comment]
cc @saad-ali

**Comment by saad-ali** [comment]
@vishh ack Will merge when safe (merged another PR that we want to monitor jenkins for before we merge more)

**Comment by AnanyaKumar** [comment]
Just got back from my vacation. Thanks @vishh and @nikhiljindal !

**Pull request merged**